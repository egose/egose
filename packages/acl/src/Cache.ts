class Cache<K, V> {
  private cache: Map<K, V>;

  constructor() {
    this.cache = new Map();
  }

  set(key: K, value: V): void {
    this.cache.set(key, value);
  }

  get(key: K): V | undefined {
    return this.cache.get(key);
  }

  delete(key: K): void {
    this.cache.delete(key);
  }

  clear(): void {
    this.cache.clear();
  }

  has(key: K): boolean {
    return this.cache.has(key);
  }

  keys(): K[] {
    return Array.from(this.cache.keys());
  }

  values(): V[] {
    return Array.from(this.cache.values());
  }

  size(): number {
    return this.cache.size;
  }
}

export default Cache;
