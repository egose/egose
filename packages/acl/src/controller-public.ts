import get from 'lodash/get';
import pick from 'lodash/pick';
import isNil from 'lodash/isNil';
import isArray from 'lodash/isArray';
import isPlainObject from 'lodash/isPlainObject';
import isBoolean from 'lodash/isBoolean';
import isString from 'lodash/isString';
import isMatch from 'lodash/isMatch';
import filter from 'lodash/filter';
import compact from 'lodash/compact';
import flatten from 'lodash/flatten';
import { normalizeSelect, iterateQuery, CustomError } from './helpers';
import Controller from './controller';
import {
  ModelRouterProps,
  MiddlewareContext,
  SubPopulate,
  ListProps,
  ReadProps,
  CreateOptionProps,
  UpdateOptionProps,
  DistinctOptionProps,
  Defaults,
  Populate,
} from './interfaces';

const filterChildren = (children, query) => {
  if (isPlainObject(query))
    return query.$and ? filter(children, (v) => query.$and.every((q) => isMatch(v, q))) : filter(children, query);

  return children;
};

const genSubPopulate = (sub: string, popul: any) => {
  if (!popul) return [];

  let populate = Array.isArray(popul) ? popul : [popul];
  populate = populate.map((p: SubPopulate | string) => {
    const ret: SubPopulate = isString(p)
      ? { path: `${sub}.${p}` }
      : {
          path: `${sub}.${p.path}`,
          select: normalizeSelect(p.select),
        };

    return ret;
  });

  return populate;
};

class PublicController extends Controller {
  async list({
    query = {},
    select = this.defaults.list?.select,
    sort = this.defaults.list?.sort,
    populate = this.defaults.list?.populate,
    limit = this.defaults.list?.limit,
    page = this.defaults.list?.page,
    options = this.defaults.list?.options || {},
  }: ListProps = {}) {
    const { includePermissions = true, includeCount = false, populateAccess = 'read', lean = false } = options;

    query = await iterateQuery(query, async (sq, key) => {
      const { model, mapper, ...rest } = sq;
      const m = this.req._macl(model);
      const arr = await m.list(rest);
      if (mapper) {
        const m = mapper.multi === false ? false : true;
        const c = mapper.compact === true;
        const f = mapper.flatten === true;
        const path = mapper.path || mapper;
        if (!m) return get(arr[0], path, isNil(mapper.defaultValue) ? null : mapper.defaultValue);

        let items = arr.map((v) => get(v, path));
        if (f) items = flatten(items);
        if (c) items = compact(items);
        return items;
      }
      return arr;
    });

    let pagination = null;
    [query, select, populate, pagination] = await Promise.all([
      this.req._genQuery(this.modelName, 'list', query),
      this.req._genSelect(this.modelName, 'list', select),
      this.req._genPopulate(this.modelName, populateAccess, populate),
      this.req._genPagination({ limit, page }, this.options.listHardLimit),
    ]);

    if (query === false) return [];

    // prevent populate paths from updating query select fields
    if (select) populate = (populate as Populate[]).filter((p) => select.includes(p.path.split('.')[0]));

    let docs = await this.model.find({ query, select, sort, populate, lean, ...pagination });
    docs = await Promise.all(
      docs.map(async (doc) => {
        if (includePermissions) doc = await this.req._permit(this.modelName, doc, 'list');
        doc = await this.req._pickAllowedFields(this.modelName, doc, 'list', ['_id', this.options.permissionField]);
        return this.req._decorate(this.modelName, doc, 'list');
      }),
    );

    const rows = await this.req._decorateAll(this.modelName, docs, 'list');

    if (includeCount) {
      return {
        count: await this.model.countDocuments(query),
        rows,
      };
    } else {
      return rows;
    }
  }

  async create(data, options: CreateOptionProps = this.defaults.create || {}) {
    const { includePermissions = true } = options as any;

    const isArr = Array.isArray(data);
    let arr = isArr ? data : [data];

    const contexts: MiddlewareContext[] = [];

    const items = await Promise.all(
      arr.map(async (item) => {
        const context: MiddlewareContext = { originalData: item };

        const allowedFields = await this.req._genAllowedFields(this.modelName, item, 'create');
        const allowedData = pick(item, allowedFields);

        const validated = await this.req._validate(this.modelName, allowedData, 'create', context);
        if (isBoolean(validated)) {
          if (!validated) throw new CustomError({ statusCode: 400, message: 'validation failed' });
        } else if (isArray(validated)) {
          if (validated.length > 0)
            throw new CustomError({ statusCode: 400, message: 'validation failed', errors: validated });
        }

        const preparedData = await this.req._prepare(this.modelName, allowedData, 'create', context);

        context.preparedData = preparedData;
        contexts.push(context);
        return preparedData;
      }),
    );

    let docs = await this.model.create(items);
    docs = await Promise.all(
      docs.map(async (doc, index) => {
        if (includePermissions) doc = await this.req._permit(this.modelName, doc, 'create', contexts[index]);
        doc = await this.req._pickAllowedFields(this.modelName, doc, 'read', ['_id', this.options.permissionField]);
        return this.req._decorate(this.modelName, doc, 'create', contexts[index]);
      }),
    );

    return isArr ? docs : docs[0];
  }

  async empty() {
    return this.model.new();
  }

  async read(
    id,
    {
      select = this.defaults.read?.select,
      populate = this.defaults.read?.populate,
      options = this.defaults.read?.options || {},
    }: ReadProps = {},
  ) {
    let access = 'read';
    const { includePermissions = true, tryList = true, populateAccess, lean = false } = options;
    const idQuery = await this.req._genIDQuery(this.modelName, id);

    let doc = await this.findById(id, {
      select,
      populate,
      options: { includePermissions, access, populateAccess, lean },
      overrides: { idQuery },
    });

    // if not found, try to get the doc with 'list' access
    if (!doc && tryList) {
      access = 'list';

      doc = await this.findById(id, {
        select,
        populate,
        options: { includePermissions, access, populateAccess, lean },
        overrides: { idQuery },
      });
    }

    if (!doc) return null;

    doc = await this.req._pickAllowedFields(this.modelName, doc, access, ['_id', this.options.permissionField]);
    doc = await this.req._decorate(this.modelName, doc, access);

    return doc;
  }

  async update(id, data, options: UpdateOptionProps = this.defaults.update || {}) {
    const { returningAll = true } = options;

    let query = await this.req._genQuery(this.modelName, 'update', await this.req._genIDQuery(this.modelName, id));
    if (query === false) return null;

    let doc = await this.model.findOne({ query });
    if (!doc) return null;

    const context: MiddlewareContext = {};

    context.originalDoc = doc.toObject();
    context.originalData = data;

    doc = await this.req._permit(this.modelName, doc, 'update', context);

    context.currentDoc = doc;
    const allowedFields = await this.req._genAllowedFields(this.modelName, doc, 'update');
    const allowedData = pick(data, allowedFields);

    const validated = await this.req._validate(this.modelName, allowedData, 'update', context);
    if (isBoolean(validated)) {
      if (!validated) throw new CustomError({ statusCode: 400, message: 'validation failed' });
    } else if (isArray(validated)) {
      if (validated.length > 0)
        throw new CustomError({ statusCode: 400, message: 'validation failed', errors: validated });
    }

    const prepared = await this.req._prepare(this.modelName, allowedData, 'update', context);

    context.preparedData = prepared;
    Object.assign(doc, prepared);

    context.modifiedPaths = doc.modifiedPaths();
    doc = await this.req._transform(this.modelName, doc, 'update', context);
    context.modifiedPaths = doc.modifiedPaths();
    doc = await doc.save();
    doc = await this.req._permit(this.modelName, doc, 'update', context);
    doc = await this.req._pickAllowedFields(this.modelName, doc, 'read', ['_id', this.options.permissionField]);
    doc = await this.req._decorate(this.modelName, doc, 'update', context, true);

    return returningAll ? doc : pick(doc, Object.keys(data));
  }

  async delete(id) {
    let query = await this.req._genQuery(this.modelName, 'delete', await this.req._genIDQuery(this.modelName, id));
    if (query === false) return null;

    let doc = await this.model.findOneAndRemove(query);
    if (!doc) return null;

    await doc.remove();
    return doc._id;
  }

  async distinct(field, options: DistinctOptionProps = this.defaults.distinct || {}) {
    let { query } = options;

    query = await this.req._genQuery(this.modelName, 'read', query);
    if (query === false) return null;

    const result = await this.model.distinct(field, query);
    if (!result) return null;

    return result;
  }

  async count(query, access = 'list') {
    query = await this.req._genQuery(this.modelName, access, query);
    if (query === false) return 0;

    return this.model.countDocuments(query);
  }

  private async getParentDoc(id, sub, access, populate = []) {
    const parentQuery = await this.req._genQuery(
      this.modelName,
      access,
      await this.req._genIDQuery(this.modelName, id),
    );

    if (parentQuery === false) return null;
    return this.model.findOne({ query: parentQuery, select: sub, populate });
  }

  async listSub(id, sub, options = {}) {
    let { filter: ft, fields } = options as any;

    const parentDoc = await this.getParentDoc(id, sub, 'read');
    if (!parentDoc) return null;
    let result = get(parentDoc, sub);

    const [subQuery, subSelect] = await Promise.all([
      this.req._genQuery(this.modelName, `subs.${sub}.list`, ft),
      this.req._genSelect(this.modelName, 'list', fields, false, [sub, 'sub']),
    ]);

    result = filterChildren(result, subQuery);
    if (subSelect) result = result.map((v) => pick(v, subSelect.concat(['id'])));
    return result;
  }

  async readSub(id, sub, subId, options = {}) {
    let { fields, populate } = options as any;

    const parentDoc = await this.getParentDoc(id, sub, 'read', genSubPopulate(sub, populate));
    if (!parentDoc) return null;
    let result = get(parentDoc, sub);

    const [subQuery, subSelect] = await Promise.all([
      this.req._genQuery(this.modelName, `subs.${sub}.read`),
      this.req._genSelect(this.modelName, 'read', fields, false, [sub, 'sub']),
    ]);

    result = filterChildren(result, subQuery);
    result = result.find((v) => String(v._id) === subId);
    if (!result) return null;

    if (subSelect) result = pick(result, subSelect.concat(['id']));
    return result;
  }

  async updateSub(id, sub, subId, data) {
    const parentDoc = await this.getParentDoc(id, sub, 'update');
    if (!parentDoc) return null;
    let result = get(parentDoc, sub);

    const [subQuery, subReadSelect, subUpdateSelect] = await Promise.all([
      this.req._genQuery(this.modelName, `subs.${sub}.update`),
      this.req._genSelect(this.modelName, 'read', null, false, [sub, 'sub']),
      this.req._genSelect(this.modelName, 'update', null, false, [sub, 'sub']),
    ]);

    result = filterChildren(result, subQuery);
    result = result.find((v) => String(v._id) === subId);
    if (!result) return null;

    const allowedData = pick(data, subUpdateSelect);
    Object.assign(result, allowedData);

    await parentDoc.save();
    if (subReadSelect) result = pick(result, subReadSelect.concat(['id']));
    return result;
  }

  async createSub(id, sub, data, options = {}) {
    let { addFirst } = options as any;

    const parentDoc = await this.getParentDoc(id, sub, 'update');
    if (!parentDoc) return null;
    let result = get(parentDoc, sub);

    const [subCreateSelect, subReadSelect] = await Promise.all([
      this.req._genSelect(this.modelName, 'create', null, false, [sub, 'sub']),
      this.req._genSelect(this.modelName, 'read', null, false, [sub, 'sub']),
    ]);

    const allowedData = pick(data, subCreateSelect);
    addFirst === true ? result.unshift(allowedData) : result.push(allowedData);

    await parentDoc.save();
    if (subReadSelect) result = result.map((v) => pick(v, subReadSelect.concat(['id'])));
    return result;
  }

  async deleteSub(id, sub, subId) {
    const parentDoc = await this.getParentDoc(id, sub, 'update');
    if (!parentDoc) return null;
    let result = get(parentDoc, sub);

    const subQuery = await this.req._genQuery(this.modelName, `subs.${sub}.delete`);

    result = filterChildren(result, subQuery);
    result = result.find((v) => String(v._id) === subId);
    if (!result) return null;

    await result.remove();
    await parentDoc.save();
    return result._id;
  }
}

export default PublicController;
