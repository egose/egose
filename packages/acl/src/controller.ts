import get from 'lodash/get';
import pick from 'lodash/pick';
import isNil from 'lodash/isNil';
import isArray from 'lodash/isArray';
import isPlainObject from 'lodash/isPlainObject';
import isBoolean from 'lodash/isBoolean';
import isString from 'lodash/isString';
import isMatch from 'lodash/isMatch';
import filter from 'lodash/filter';
import compact from 'lodash/compact';
import flatten from 'lodash/flatten';
import Model from './model';
import { setModelOptions, setModelOption, getModelOptions } from './options';
import { normalizeSelect, iterateQuery, CustomError } from './helpers';
import {
  ModelRouterProps,
  MiddlewareContext,
  SubPopulate,
  ListProps,
  ReadProps,
  CreateOptionProps,
  UpdateOptionProps,
  DistinctOptionProps,
  Defaults,
  Populate,
  FindOneOptionProps,
  FindOneProps,
  FindByIdProps,
} from './interfaces';

class Controller {
  req: any;
  modelName: string;
  model: Model;
  options: ModelRouterProps;
  defaults: Defaults;

  constructor(req: any, modelName: string) {
    this.req = req;
    this.modelName = modelName;
    this.model = new Model(modelName);
    this.options = getModelOptions(modelName);
    this.defaults = this.options.defaults || {};
  }

  protected async findOne({
    query: _query = {},
    select: _select = this.defaults.findOne?.select,
    populate: _populate = this.defaults.findOne?.populate,
    options = this.defaults.findOne?.options || {},
    overrides = {},
  }: FindOneProps = {}) {
    const { includePermissions = true, access = 'read', populateAccess, lean = false } = options;
    const { query: __query, select: __select, populate: __populate } = overrides;

    let [query, select, populate] = await Promise.all([
      __query || this.req._genQuery(this.modelName, access, _query),
      __select || this.req._genSelect(this.modelName, access, _select),
      __populate || this.req._genPopulate(this.modelName, populateAccess || access, _populate),
    ]);

    if (query === false) return null;

    let doc = await this.model.findOne({ query, select, populate, lean });
    if (!doc) return null;

    if (includePermissions) doc = await this.req._permit(this.modelName, doc, access);
    return doc;
  }

  protected async findById(
    id,
    {
      select: _select = this.defaults.findOne?.select,
      populate: _populate = this.defaults.findOne?.populate,
      options = this.defaults.findOne?.options || {},
      overrides = {},
    }: FindByIdProps = {},
  ) {
    const { select: __select, populate: __populate, idQuery: __idQuery } = overrides;
    const query = __idQuery || (await this.req._genIDQuery(this.modelName, id));

    return this.findOne({
      query,
      select: _select,
      populate: _populate,
      options,
      overrides: {
        select: __select,
        populate: __populate,
      },
    });
  }
}

export default Controller;
