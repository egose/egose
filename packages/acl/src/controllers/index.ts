export * from './controller';
export * from './public-controller';
