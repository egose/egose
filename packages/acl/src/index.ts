import { Request, Response, NextFunction } from 'express';
import isNil from 'lodash/isNil';
import isUndefined from 'lodash/isUndefined';
import middleware from './middleware';
import { RootRouter, ModelRouter } from './router';
import { setGlobalOption, setGlobalOptions, getGlobalOption } from './options';
import { GlobalOptions, RootRouterOptions, ModelRouterOptions } from './interfaces';
export * from './router';
export * from './options';
export * from './interfaces';
export * from './symbols';

type Middleware = () => (req: Request, res: Response, next: NextFunction) => Promise<void>;
interface ModelRouterConstructor {
  new (modelName: string, options: ModelRouterOptions): ModelRouter;
}

type CreateRouter = {
  (modelName: string, options: ModelRouterOptions): ModelRouter;
  (options: RootRouterOptions): RootRouter;
};

interface MACL {
  createRouter: CreateRouter;
  set: (keyOrOptions: string | GlobalOptions, value?: any) => void;
  setGlobalOption: (optionKey: string, value: any) => void;
  setGlobalOptions: (options: GlobalOptions) => void;
  getGlobalOption: (optionKey: string, defaultValue?: any) => any;
  ModelRouter: ModelRouterConstructor;
}

const macl = middleware as Middleware & MACL;
macl.createRouter = function (modelName: string | RootRouterOptions, options: ModelRouterOptions | undefined) {
  return isUndefined(options)
    ? new RootRouter(modelName as RootRouterOptions)
    : new ModelRouter(modelName as string, options);
} as CreateRouter;

macl.set = (keyOrOptions, value) =>
  isNil(value) ? setGlobalOptions(keyOrOptions as GlobalOptions) : setGlobalOption(keyOrOptions as string, value);
macl.setGlobalOption = setGlobalOption;
macl.setGlobalOptions = setGlobalOptions;
macl.getGlobalOption = getGlobalOption;
macl.ModelRouter = ModelRouter;

export default macl;
