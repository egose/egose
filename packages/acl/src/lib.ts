import Document from 'mongoose/lib/document';
import Schema from 'mongoose/lib/schema';
import isFunction from 'lodash/isFunction';
import isPlainObject from 'lodash/isPlainObject';

export const isSchema = (val) => val instanceof Schema;
export const isObjectIdType = (val) => val === 'ObjectId' || val === Schema.Types.ObjectId;
export const isReference = (val) => isPlainObject(val) && val.ref && isObjectIdType(val.type);

export const isPromise = function isPromise(val) {
  return val && val.then && isFunction(val.then);
};

export const isDocument = function isDocument(doc) {
  return doc instanceof Document;
};

export const toAsyncFn = function toAsyncFn(fn: Function, defaultValue?: any) {
  if (!fn) return () => Promise.resolve(defaultValue);
  return function asyncFn(...args) {
    const ret = fn.apply(this, args);
    return isPromise(ret) ? ret : Promise.resolve(ret);
  };
};
