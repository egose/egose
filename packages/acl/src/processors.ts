import castArray from 'lodash/castArray';
import cloneDeep from 'lodash/cloneDeep';
import forEach from 'lodash/forEach';
import get from 'lodash/get';
import isArray from 'lodash/isArray';
import map from 'lodash/map';
import set from 'lodash/set';

interface ProcessCopy {
  src: string;
  dest: string;
}

interface CopyAndDepopulateOptions {
  mutable?: boolean;
  idField?: string;
}

export const copyAndDepopulate = (
  docObject: any,
  operations: ProcessCopy[],
  options: CopyAndDepopulateOptions = { mutable: true, idField: '_id' },
) => {
  const obj = get(options, 'mutable', true) ? docObject : cloneDeep(docObject);
  const idField = get(options, 'idField', '_id');

  forEach(castArray<ProcessCopy>(operations), (op) => {
    if (!op.src || !op.dest) return;

    let targets = [obj];
    const segs = op.src.split('.');
    forEach<string>(segs, (seg, ind) => {
      if (segs.length === ind + 1) {
        forEach(targets, (target) => {
          set(target, op.dest, get(target, seg));
          set(target, seg, isArray(target[seg]) ? map(target[seg], idField) : get(target, `${seg}.${idField}`));
        });
      } else {
        targets = targets.reduce((ret, target) => {
          if (isArray(target[seg])) ret.push(...target[seg]);
          else ret.push(target[seg]);

          return ret;
        }, []);
      }
    });
  });

  return obj;
};
