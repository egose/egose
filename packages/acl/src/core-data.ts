import { Response, NextFunction } from 'express';
import mongoose, { Model } from 'mongoose';
import assign from 'lodash/assign';
import castArray from 'lodash/castArray';
import compact from 'lodash/compact';
import difference from 'lodash/difference';
import forEach from 'lodash/forEach';
import get from 'lodash/get';
import intersection from 'lodash/intersection';
import isArray from 'lodash/isArray';
import isBoolean from 'lodash/isBoolean';
import isEmpty from 'lodash/isEmpty';
import isFunction from 'lodash/isFunction';
import isNaN from 'lodash/isNaN';
import isPlainObject from 'lodash/isPlainObject';
import isString from 'lodash/isString';
import isNil from 'lodash/isNil';
import noop from 'lodash/noop';
import pick from 'lodash/pick';
import set from 'lodash/set';
import reduce from 'lodash/reduce';
import { getGlobalOption, getDataOption } from './options';
import {
  Populate,
  Projection,
  Filter,
  MiddlewareContext,
  Validation,
  Request,
  SelectAccess,
  RouteGuardAccess,
  DocPermissionsAccess,
  BaseFilterAccess,
  DecorateAccess,
  DecorateAllAccess,
  ValidateAccess,
  PrepareAccess,
  TransformAccess,
  Task,
} from './interfaces';
import Permission, { Permissions } from './permission';
import { DataService } from './services';
import { normalizeSelect, createValidator, setDocValue, toObject, pickDocFields, genPagination } from './helpers';
import { copyAndDepopulate } from './processors';
import { isDocument, arrToObj } from './lib';
import { DATA_MIDDLEWARE, PERMISSIONS, PERMISSION_KEYS } from './symbols';
import { Cache } from './cache';

export class DataCore {
  private req: Request;
  private caches: {
    baseFilter: Cache<string, any>;
  };

  constructor(req: Request) {
    this.req = req;
    this.caches = {
      baseFilter: new Cache<string, any>(),
    };
  }

  async genIDFilter(dataName: string, id: string) {
    const identifier = getDataOption(dataName, 'identifier');

    if (isString(identifier)) {
      return { [identifier]: id };
    }

    if (isFunction(identifier)) {
      return identifier.call(this.req, id);
    }

    return { _id: id };
  }

  async genFilter(dataName: string, access: BaseFilterAccess = 'read', _filter: Filter = null): Promise<Filter> {
    const baseFilterFn = getDataOption(dataName, `baseFilter.${access}`, null);
    if (!isFunction(baseFilterFn)) return _filter || {};

    const cacheKey = `${dataName}_baseFilter_${access}`;
    if (this.caches.baseFilter.has(cacheKey)) {
      return this.caches.baseFilter.get(cacheKey);
    }

    const permissions = this.getGlobalPermissions();

    const baseFilter = await baseFilterFn.call(this.req, permissions);
    if (baseFilter === false) return false;
    if (baseFilter === true || isEmpty(baseFilter)) return _filter || {};
    if (!_filter) return baseFilter;

    const result = { $and: [baseFilter, _filter] };
    this.caches.baseFilter.set(cacheKey, result);
    return result;
  }

  async genAllowedFields(dataName: string, doc: any, access: SelectAccess, baseFields = []) {
    let fields = [...baseFields] || [];

    const permissionSchema = getDataOption(dataName, 'permissionSchema');
    if (!permissionSchema) return fields;

    const permissions = this.getGlobalPermissions();

    // get keys from permission schema as some fields might not be filled when created
    const keys = Object.keys(permissionSchema);

    const phas = (key) => permissions.has(key);
    const { stringHandler, arrayHandler } = createValidator(phas);

    for (let x = 0; x < keys.length; x++) {
      const key = keys[x];
      if (baseFields.includes(key)) continue;

      const val = permissionSchema[key];
      const value = (val && val[access]) || val;

      if (isBoolean(value)) {
        if (value) fields.push(key);
      } else if (isString(value)) {
        if (stringHandler(value)) fields.push(key);
      } else if (isArray(value)) {
        if (arrayHandler(value)) fields.push(key);
      } else if (isFunction(value)) {
        if (await value.call(this.req, permissions)) fields.push(key);
      }
    }

    return fields;
  }

  async pickAllowedFields(dataName: string, doc: any, access: SelectAccess, baseFields = []) {
    const allowed = await this.genAllowedFields(dataName, doc, access, baseFields);
    return pickDocFields(doc, allowed);
  }

  async genSelect(
    dataName: string,
    access: SelectAccess,
    targetFields: Projection = null,
    skipChecks = true,
    subPaths = [],
  ) {
    let normalizedSelect = normalizeSelect(targetFields);
    let fields = [];

    const permissionSchema = getDataOption(dataName, ['permissionSchema'].concat(subPaths).join('.'));
    if (!permissionSchema) return fields;

    const permissions = this.getGlobalPermissions();

    const phas = (key) => {
      if (permissions.prop(key)) {
        if (permissions.has(key)) return true;
      } else if (skipChecks) {
        return true;
      }

      return false;
    };

    const { stringHandler, arrayHandler } = createValidator(phas);

    const keys = Object.keys(permissionSchema);
    for (let x = 0; x < keys.length; x++) {
      const key = keys[x];
      const val = permissionSchema[key];
      const value = val[access] || val;

      if (isBoolean(value)) {
        if (value) fields.push(key);
      } else if (isString(value)) {
        if (stringHandler(value)) fields.push(key);
      } else if (isArray(value)) {
        if (arrayHandler(value)) fields.push(key);
      } else if (isFunction(value)) {
        if (await value.call(this.req, permissions)) fields.push(key);
      }
    }

    fields = intersection(normalizedSelect, fields);
    return fields;
  }

  async decorate(dataName: string, doc: any, access: DecorateAccess, context: MiddlewareContext = {}) {
    const decorate = getDataOption(dataName, `decorate.${access}`, null);

    const permissions = this.getGlobalPermissions();
    return this.callMiddleware(decorate, doc, permissions, context);
  }

  async decorateAll(dataName: string, docs: any[], access: DecorateAllAccess) {
    const decorateAll = getDataOption(dataName, `decorateAll.${access}`, null);
    const permissions = this.getGlobalPermissions();

    return this.callMiddleware(decorateAll, docs, permissions, {});
  }

  getPermissions() {
    const permissionField = getGlobalOption('permissionField');
    return new Permission(this.req[permissionField] || {});
  }

  async setPermissions() {
    const permissionField = getGlobalOption('permissionField');
    if (this.req[permissionField]) return;

    const globalPermissions = getGlobalOption('globalPermissions');
    if (isFunction(globalPermissions)) {
      const gp = await globalPermissions.call(this.req, this.req);
      if (isPlainObject(gp)) this.req[permissionField] = gp;
      else if (isArray(gp)) this.req[permissionField] = arrToObj(gp);
      else if (isString(gp)) this.req[permissionField] = { [gp]: true };
    }
  }

  async canActivate(routeGuard: Validation) {
    let allowed = false;

    const permissions = this.getGlobalPermissions();
    const phas = (key) => permissions.has(key);
    const { stringHandler, arrayHandler } = createValidator(phas);

    if (isBoolean(routeGuard)) {
      allowed = routeGuard === true;
    } else if (isString(routeGuard)) {
      allowed = stringHandler(routeGuard);
    } else if (isArray(routeGuard)) {
      allowed = arrayHandler(routeGuard);
    } else if (isFunction(routeGuard)) {
      allowed = routeGuard.call(this.req, permissions);
    }

    return allowed;
  }

  async isAllowed(dataName: string, access: RouteGuardAccess | string) {
    const routeGuard = getDataOption(dataName, `routeGuard.${access}`);
    return this.canActivate(routeGuard);
  }

  getService(dataName: string) {
    return new DataService(this.req, dataName);
  }

  service(dataName: string) {
    return this.getService(dataName);
  }

  svc(dataName: string) {
    return this.getService(dataName);
  }

  private getGlobalPermissions() {
    return this.req[PERMISSIONS] as Permission;
  }

  private async callMiddleware(
    middleware: Function | Function[],
    doc: any,
    permissions: Permissions,
    context: MiddlewareContext,
  ) {
    middleware = castArray(middleware);
    for (let x = 0; x < middleware.length; x++) {
      if (isFunction(middleware[x])) {
        doc = await middleware[x].call(this.req, doc, permissions, context);
      }
    }

    return doc;
  }
}

export async function setDataCore(req: Request, res: Response, next: NextFunction) {
  if (req[DATA_MIDDLEWARE]) return next();

  const core = new DataCore(req);
  await core.setPermissions();

  req.dacl = core;
  req[PERMISSIONS] = core.getPermissions();
  req[PERMISSION_KEYS] = req[PERMISSIONS].$_permissionKeys;
  req[DATA_MIDDLEWARE] = true;

  next();
}
