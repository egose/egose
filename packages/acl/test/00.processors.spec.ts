import 'mocha';
import chai, { expect } from 'chai';
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot';
import { copyAndDepopulate } from '@egose/acl/src/processors';

chai.use(jestSnapshotPlugin());

describe('Processors - copyAndDepopulate', () => {
  it('should return the expected result - 1', async () => {
    const data = { apple: { _id: 'qwer', name: 'apple' }, pear: { _id: 'asdf', name: 'pear' } };
    const result = copyAndDepopulate(data, [{ src: 'apple', dest: '_apple' }]);
    expect(result).toMatchSnapshot();
  });

  it('should return the expected result - 2', async () => {
    const data = {
      apple: { _id: 'qwer', name: 'apple' },
      pear: {
        _id: 'asdf',
        items: [
          { _id: 1, name: 'item1' },
          { _id: 2, name: 'item2' },
        ],
      },
    };
    const result = copyAndDepopulate(data, [{ src: 'pear.items', dest: '_items' }], { mutable: false });
    expect(result).to.be.not.equal(data);
    expect(result).toMatchSnapshot();
  });

  it('should return the expected result - 3', async () => {
    const data = {
      apple: { _id: 'qwer', name: 'apple' },
      pear: {
        _id: 'asdf',
        items: [
          {
            _id: 1,
            name: 'item1',
            samples: [
              { _id: 1, name: 'sample1' },
              { _id: 2, name: 'sample2' },
            ],
          },
          {
            _id: 2,
            name: 'item2',
            samples: [
              { _id: 3, name: 'sample3' },
              { _id: 4, name: 'sample4' },
            ],
          },
        ],
      },
    };

    const result = copyAndDepopulate(data, [{ src: 'pear.items.samples', dest: '_samples' }], { idField: 'name' });
    expect(result).to.be.equal(data);
    expect(result).toMatchSnapshot();
  });
});
