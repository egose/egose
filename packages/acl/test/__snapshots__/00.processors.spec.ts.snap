// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processors - copyAndDepopulate should return the expected result - 1 1`] = `
Object {
  "_apple": Object {
    "_id": "qwer",
    "name": "apple",
  },
  "apple": "qwer",
  "pear": Object {
    "_id": "asdf",
    "name": "pear",
  },
}
`;

exports[`Processors - copyAndDepopulate should return the expected result - 2 1`] = `
Object {
  "apple": Object {
    "_id": "qwer",
    "name": "apple",
  },
  "pear": Object {
    "_id": "asdf",
    "_items": Array [
      Object {
        "_id": 1,
        "name": "item1",
      },
      Object {
        "_id": 2,
        "name": "item2",
      },
    ],
    "items": Array [
      1,
      2,
    ],
  },
}
`;

exports[`Processors - copyAndDepopulate should return the expected result - 3 1`] = `
Object {
  "apple": Object {
    "_id": "qwer",
    "name": "apple",
  },
  "pear": Object {
    "_id": "asdf",
    "items": Array [
      Object {
        "_id": 1,
        "_samples": Array [
          Object {
            "_id": 1,
            "name": "sample1",
          },
          Object {
            "_id": 2,
            "name": "sample2",
          },
        ],
        "name": "item1",
        "samples": Array [
          "sample1",
          "sample2",
        ],
      },
      Object {
        "_id": 2,
        "_samples": Array [
          Object {
            "_id": 3,
            "name": "sample3",
          },
          Object {
            "_id": 4,
            "name": "sample4",
          },
        ],
        "name": "item2",
        "samples": Array [
          "sample3",
          "sample4",
        ],
      },
    ],
  },
}
`;
